Do `io.BytesIO()` instances need closing? I thought that they were just a mock file-like wrapper around a string but this seems to say otherwise?
```python
/e/n/pyinstaller (pyup/scheduled-update-2022-03-06)> pytest -W 'error::ResourceWarning' -k basic tests/functional/test_basic.py::test_run_from_path_environ
============================================================================================================ test session starts =============================================================================================================
platform linux -- Python 3.9.9, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /home/brenainn/.pyenv/versions/3.9.9/bin/python3.9
cachedir: .pytest_cache
rootdir: /e/notebooks/pyinstaller, configfile: setup.cfg
plugins: rerunfailures-10.2, stress-1.0.1, drop-dup-tests-0.3.0, cookies-0.6.1, flaky-3.7.0, order-1.0.1, forked-1.4.0, xdist-2.5.0, timeout-2.1.0, cov-3.0.0
timeout: 600.0s
timeout method: signal
timeout func_only: False
collected 2 items                                                                                                                                                                                                                            

tests/functional/test_basic.py::test_run_from_path_environ[onedir] PASSED                                                                                                                                                              [ 50%]
tests/functional/test_basic.py::test_run_from_path_environ[onefile] PASSED                                                                                                                                                             [100%]

============================================================================================================== warnings summary ==============================================================================================================
PyInstaller/lib/modulegraph/modulegraph.py:20
  /e/notebooks/pyinstaller/PyInstaller/lib/modulegraph/modulegraph.py:20: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

tests/functional/test_basic.py::test_run_from_path_environ[onedir]
tests/functional/test_basic.py::test_run_from_path_environ[onefile]
  /home/brenainn/.pyenv/versions/3.9.9/lib/python3.9/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning: Exception ignored in: <_io.FileIO [closed]>
  
  Traceback (most recent call last):
    File "/e/notebooks/pyinstaller/PyInstaller/archive/writers.py", line 377, in add
      fh = io.BytesIO(data)
  ResourceWarning: unclosed file <_io.BufferedReader name='/e/notebooks/pyinstaller/PyInstaller/loader/pyimod04_ctypes.pyc'>
  
    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
======================================================================================================= 2 passed, 3 warnings in 19.76s =======================================================================================================
```

--- a/PyInstaller/building/build_main.py
+++ b/PyInstaller/building/build_main.py
@@ -262,6 +262,14 @@ class Analysis(Target):
         logger.info('Extending PYTHONPATH with paths\n' + pprint.pformat(self.pathex))
         sys.path.extend(self.pathex)
 
+        # If pkg_resources has already been imported, force update of its working set to account for changes made to
+        # sys.path. Otherwise, distribution data in the added path(s) may not be discovered.
+        if 'pkg_resources' in sys.modules:
+            # See https://github.com/pypa/setuptools/issues/373
+            import pkg_resources
+            if hasattr(pkg_resources, '_initialize_master_working_set'):
+                pkg_resources._initialize_master_working_set()
+
         # Set global variable to hold assembly binding redirects
         CONF['binding_redirects'] = []
 

--- a/PyInstaller/utils/hooks/__init__.py
+++ b/PyInstaller/utils/hooks/__init__.py
@@ -889,9 +889,22 @@ def copy_metadata(package_name, recursive=False):
         package_name = todo.pop()
         if package_name in done:
             continue
+
         dist = pkg_resources.get_distribution(package_name)
-        dest = _copy_metadata_dest(dist.egg_info, dist.project_name)
-        out.append((dist.egg_info, dest))
+        if dist.egg_info is not None:
+            # If available, dist.egg_info points to the source .egg-info or .dist-info directory.
+            dest = _copy_metadata_dest(dist.egg_info, dist.project_name)
+            out.append((dist.egg_info, dest))
+        else:
+            # When .egg-info is not a directory but a single file, dist.egg_info is None, and we need to resolve the
+            # path ourselves. This format is common on Ubuntu/Debian with their deb-packaged python packages.
+            dist_src = _resolve_legacy_metadata_path(dist)
+            if dist_src is None:
+                raise RuntimeError(
+                    f"No metadata path found for distribution '{dist.project_name}' (legacy fallback search failed)."
+                )
+            out.append((dist_src, '.'))  # It is a file, so dest path needs to be '.'
+
         if not recursive:
             return out
         done.add(package_name)
@@ -902,6 +915,41 @@ def copy_metadata(package_name, recursive=False):
 
 def _normalise_dist(name: str) -> str:
     return name.lower().replace("_", "-")
+
+
+def _resolve_legacy_metadata_path(dist):
+    """
+    Attempt to resolve the legacy metadata file for the given distribution.
+    The .egg-info file is commonly used by Debian/Ubuntu when packaging python packages.
+
+    Args:
+        dist:
+            The distribution information as returned by ``pkg_resources.get_distribution("xyz")``.
+    Returns:
+        The path to the distribution's metadata file.
+    """
+
+    candidates = [
+        # This fallback was in place in pre-#5774 times. However, it is insufficient, because dist.egg_name() may be
+        # greenlet-0.4.15-py3.8 (Ubuntu 20.04 package) while the file we are searching for is greenlet-0.4.15.egg-info.
+        f"{dist.egg_name()}.egg-info",
+        # The extra name-version.egg-info path format
+        f"{dist.project_name}-{dist.version}.egg-info",
+        # And the name_with_underscores-version.egg-info.format
+        f"{dist.project_name.replace('-', '_')}-{dist.version}.egg-info",
+    ]
+
+    # As an additional attempt, try to remove the-pyX.Y suffix from egg name.
+    pyxx_suffix = f"-py{sys.version_info[0]}.{sys.version_info[1]}"
+    if dist.egg_name().endswith(pyxx_suffix):
+        candidates.append(dist.egg_name()[:-len(pyxx_suffix)] + ".egg-info")
+
+    for candidate in candidates:
+        candidate_path = os.path.join(dist.location, candidate)
+        if os.path.isfile(candidate_path):
+            return candidate_path
+
+    return None
 
 
 def _copy_metadata_dest(egg_path: str, project_name: str) -> str:

--- a/news/6692.bugfix.rst
+++ b/news/6692.bugfix.rst
@@ -1,0 +1,4 @@
+Fix collection of package metadata when the latter is provided in the
+form of ``.egg-info`` file instead of more conventional ``.egg-info``
+directory or standard ``.dist-info`` directory. Such metadata format
+is often found in Debian and Ubuntu deb packages of python software.

--- a/tests/functional/modules/pyi_single_file_metadata/README
+++ b/tests/functional/modules/pyi_single_file_metadata/README
@@ -1,0 +1,4 @@
+Package metadata stored in a single-file .egg-info format, as commonly
+found in Debian/Ubuntu packages. Used by test_misc::test_single_file_metadata
+which adds this directory to its search path.
+

--- a/tests/functional/modules/pyi_single_file_metadata/my_test_package-1.0.egg-info
+++ b/tests/functional/modules/pyi_single_file_metadata/my_test_package-1.0.egg-info
@@ -1,0 +1,14 @@
+Metadata-Version: 2.1
+Name: my-test-package
+Version: 1.0
+Summary: Test package for single-file metadata
+Author: PyInstaller Team
+License: MIT
+Platform: Windows
+Platform: Linux
+Platform: Solaris
+Platform: Mac OS-X
+Platform: Unix
+Requires-Python: >=3.6
+
+The package for single-file metadata.

--- a/tests/functional/test_misc.py
+++ b/tests/functional/test_misc.py
@@ -132,3 +132,25 @@ def test_onefile_cleanup_symlinked_dir(pyi_builder, tmpdir):
     for idx in range(5):
         output_file = os.path.join(output_dir, f'output-{idx}.txt')
         assert os.path.isfile(output_file)
+
+
+# Test that single-file metadata (as commonly found in Debian/Ubuntu packages) is properly collected by copy_metadata().
+def test_single_file_metadata(pyi_builder):
+    # Add directory containing the my-test-package metadata to search path
+    extra_path = os.path.join(_MODULES_DIR, "pyi_single_file_metadata")
+
+    pyi_builder.test_source(
+        """
+        import pkg_resources
+
+        # The pkg_resources.get_distribution() call automatically triggers collection of the metadata. While it does not
+        # raise an error if metadata is not found while freezing, the calls below will fall at run-time in that case.
+        dist = pkg_resources.get_distribution('my-test-package')
+
+        # Sanity check
+        assert dist.project_name == 'my-test-package'
+        assert dist.version == '1.0'
+        assert dist.egg_name() == f'my_test_package-{dist.version}-py{sys.version_info[0]}.{sys.version_info[1]}'
+        """,
+        pyi_args=['--paths', extra_path]
+    )

